[jade.xyz] 방탈출 인증 관리 step1,2 과제 제출

mincastle98 opened the pull request:
안녕하세요 제이든! 이번에도 페이로 진행되는 과제 때문에 새로운 과제 1차 리뷰부터 요청드립니다.
이번 과제는 지난 과제에서 인증/인가 기능이 추가되었습니다.
기능은 지난 주에 진행한 과제와 동일하지만 이번 과제에서는 주어진 뼈대 코드 활용해서 진행했습니다.
auth 디렉토리 위주로 확인해주시면 감사하겠습니다!
테스트 코드는 페어와 다른 방식으로 작성하여서 `RestAssured`와 `Mockmvc`가 섞여 있습니다.

확인하시고 리뷰 부탁드립니다. 감사합니다!

gracefulBrown committed:
default code

gracefulBrown committed:
console app

gracefulBrown committed:
boot & mvc

gracefulBrown committed:
db

gracefulBrown committed:
refactor(reservation): layered

gracefulBrown committed:
refactor(reservation): exception

gracefulBrown committed:
test(reservation): add e2e test

boorownie committed:
refactor(reservation)

boorownie committed:
feat(theme)

boorownie committed:
feat(schedule)

boorownie committed:
fix(test)

boorownie committed:
feat(member)

boorownie committed:
feat(token): create token

boorownie committed:
Initial commit

boorownie committed:
feat(token): create token

gusah009 committed:
docs: step1 기능 요구사항 작성

gusah009 committed:
fix: request dto 기본 생성자 추가

gusah009 committed:
style: formatting

gusah009 committed:
build: spring-boot-starter-validation 추가

gusah009 committed:
feat: 토큰 발급하는 API 추가

gusah009 committed:
test: 잘못된 클래스명 수정

ThemeE2ETest -> MemberE2ETest

gusah009 committed:
feat: 내 정보 조회하기 API 추가

gusah009 committed:
docs: step2 기능 요구사항 작성

gusah009 committed:
feat: 예약 생성 API 추가

gusah009 committed:
feat: 예약 취소 API 추가

gusah009 committed:
refactor: 비밀번호 체크 로직 member로 이동

jayden-17 commented on README.md:
```diff
@@ -0,0 +1,31 @@
+# Step1
```
요구사항 정리 👍

jayden-17 commented on src/main/java/nextstep/auth/TokenRequest.java:
```diff
@@ -1,9 +1,17 @@
+    @NotNull
     private String password;
 
+    private TokenRequest() {
```
private 기본 생성자는 어떤 목적으로 추가되었을까요 ?

mincastle98 replied:
Reservation 테스트 코드에서 기존 상태에서는 오류가 발생했었고 기본 생성자를 추가하면서 오류가 해결되어 다른 RequestDto에 모두 기본 생성자를 추가했었습니다. 찾아보니 @RequestBody 어노테이션 동작 중 `Jackson2HttpMessageConverter`에서 매핑이 이뤄지는데 이때 기본 생성자가 필요하다고 하네요. 그런데 의문은 다른 Dto에서는 기본 생성자를 제거해도 정상적으로 동작이 되는데 reservationRequestDto에서만 오류가 발생합니다. 혹시 제가 잘못 이해한 부분이 있을까요?

jayden-17 replied:
`TokenRequest` 는  `TokenRequest(String username, String password)` 생성자를 통해 값을 할당하고 getter 를 통해 값을 조회하도록 설계되어 있습니다. 불필요하게(framework 스펙에 의해) 기본 생성자가 추가되어 의도하지 않는 방식으로 객체가 사용될 우려가 있습니다. 다음과 같은 불변 객체로 설계된 DTO 를 framework 에서 사용할 방법이 있는지 확인해보면 좋을 것 같습니다.

---

객체를 생성하고 값을 할당하는 방법들은 여러가지가 있을 것 같습니다. 

1. 기본 생성자로 생성 후 setter 로 값을 할당 (POJO)
2. 생성자를 통해 값을 할당 (불변 객체 사용을 위한)
4. reflection 을 통해 값을 할당 (서비스 로직에선 일반적이지 않음)

객체가 일관성 있게만 처리되면 좋을 것 같습니다.
1 or 2 번으로 객체를 설계 했다면 프레임워크에서 해당 방법으로 생성되게 유도하면 좋을 것 같아요 ~ 

--- 

제가 `ReservationRequest` 의 기본 생성자를 제거하고 `ReservationControllerTest` 실행시 테스트가 통과하는 것 같은데 혹시 오류가 발생하는 케이스를 알려주시면 같이 확인해볼게요 ~

jayden-17 commented on src/main/java/nextstep/auth/AuthController.java:
```diff
@@ -0,0 +1,24 @@
+@RestController
+public class AuthController {
+
+    private AuthService authService;
```
어떤 곳은 `private final` 로 선언되어 있고 
어떤 곳은 `private` 으로만 선언되어 있는데 
쓰이는 목적에 맞게 통일하는게 어떨까요 ?

jayden-17 commented on src/main/java/nextstep/reservation/ReservationController.java:
```diff
@@ -17,8 +18,9 @@ public ReservationController(ReservationService reservationService) {
-        Long id = reservationService.create(reservationRequest);
+    public ResponseEntity createReservation(@AuthenticationPrincipal Long authId,
+                                            @RequestBody ReservationRequest reservationRequest) {
+        Long id = reservationService.create(authId, reservationRequest);
```
reservation 관점에서는 authId 보다는 memberId 나 userId 또는 reservationUserId 등이 더 적절하지 않을까요 ?

jayden-17 commented on src/main/java/nextstep/reservation/ReservationService.java:
```diff
@@ -49,12 +56,15 @@ public List<Reservation> findAllByThemeIdAndDate(Long themeId, String date) {
+    public void deleteById(Long authId, Long reservationId) {
+        Reservation reservation = reservationDao.findById(reservationId);
         if (reservation == null) {
             throw new NullPointerException();
```
NullPointerException 보다 더 적절한 예외처리가 가능할 것 같습니다.

jayden-17 commented on src/main/java/nextstep/reservation/ReservationService.java:
```diff
@@ -49,12 +56,15 @@ public List<Reservation> findAllByThemeIdAndDate(Long themeId, String date) {
-        reservationDao.deleteById(id);
+        Member me = memberDao.findById(authId);
+        if (!reservation.isMyReservation(me)) {
+            throw new UnAuthorizationException();
```
서비스 로직에서 봤을 때 인증 에러 보다는 
접근할 수 없는 예약 정보 등 다른 예외를 정의해보면 어떨까요 ?

jayden-17 reviewed (changes requested):
매번 요구 사항도 꼼꼼하게 잘 정리해주셔서 리뷰하는데 많은 도움이 되고 있습니다 👍 
요청 주신데로 `auth` 패키지 위주로 몇가지 질문과 커맨트 남겨두었습니다. 

확인 후 다시 요청 부탁드려요 😄

kakao-jade-xyz committed:
refactor: 미사용 메서드 삭제

kakao-jade-xyz committed:
refactor: RestAssured response 타입 수정

kakao-jade-xyz committed:
fix: Test 코드 및 ddl 오류 일부 수정

kakao-jade-xyz committed:
refactor: 불필요한 기본 생성자 제거

kakao-jade-xyz committed:
refactor: final 추가

- 뼈대코드를 바탕으로 작성한 코드여서 추가 작성한 코드에 대해서만 적용하였습니다.

kakao-jade-xyz committed:
refactor: Principal 변수명 수정

- 변수명 authId를 객체의 관점에 맞춰 userId로 수정

kakao-jade-xyz committed:
refactor: exception 패키지 이동

- Custom Exception을 exception 패키지에서 관리하도록 이동

kakao-jade-xyz committed:
fix: 권한없는 예약 접근 시 발생하는 Exception 수정

- 권한이 없는 예약 정보에 접근 시 UnAuthorizationException이 아닌 InaccessibleReservationException이 발생하도록 수정

kakao-jade-xyz committed:
refactor: 뼈대 코드 일부 수정

- reservation 패키지 내 뼈대 코드 일부 수정

mincastle98 commented:
뼈대 코드를 그대로 사용하면서 부족했던 부분 보완하였고 피드백 남겨주신 내용 반영하였습니다.
추가로 궁금한 점 코멘트에 남겨두었습니다.

감사합니다🙇🏻‍♂️

jayden-17 reviewed (approved):
리뷰한 부분 잘 반영해주셨네요 ~ 고생하셨습니다 👍 

이번 PR 을 merge 하도록 할게요 ~ 

질문 주신 부분도 답변 남겨 놓았습니다.
